### PR TITLE
Adds checkpointing saving and loading

### PR DESCRIPTION
**Pull request by edbeeching**
This PR adds the saving and loading of the model, optimizer and scheduler states.
I don't think it is really needed to store the ref model weights, but it was not clear how to get accelerate.save_state to ignore this.

I updated the example gpt2-sentiment to demonstrate the usage, we will probably want to revert these changes before merging.
I have not tested with a peft model yet.

Early feedback would be great @younesbelkada

### CODE REVIEW

File: trl/trainer/ppo_trainer.py [hunk 1]
```diff
@@ -589,6 +622,10 @@ def collator(data):
         if self.lr_scheduler is not None:
             self.lr_scheduler.step()
 
+        self.current_step += 1
+        if self.config.save_steps and self.current_step % self.config.save_steps == 0:
+            self.accelerator.save_state()
```
Comment (younesbelkada) [inline trl/trainer/ppo_trainer.py:627]: Nice! Does this needs to be executed on the main rank in case of distributed training? Maybe the `accelerator` already takes care of that

File: trl/models/modeling_value_head.py [hunk 2]
```diff
@@ -200,15 +200,15 @@ def state_dict(self, *args, **kwargs):
         to the state dictionary of the wrapped model by prepending the key with `v_head.`.
         """
         if not self.is_peft_model:
-            pretrained_model_state_dict = self.pretrained_model.state_dict(*args, **kwargs)
-        else:
-            # if it is a peft model, only save the v_head
-            pretrained_model_state_dict = {}
+            return super(PreTrainedModelWrapper, self).state_dict(*args, **kwargs)
```
Comment (younesbelkada) [inline trl/models/modeling_value_head.py:203]:
```suggestion
            state_dict = super(PreTrainedModelWrapper, self).state_dict(*args, **kwargs)
            v_head_state_dict = self.v_head.state_dict(*args, **kwargs)
            for k, v ...
```

**Review by younesbelkada — COMMENTED** [review]
Looks great! 
I left some suggestions to make the tests pass
Can you also add a test to check that after saving the trainer things work as expected?
Thanks!

File: trl/models/modeling_value_head.py (hunk 2 above)
Comment (younesbelkada) [reply trl/models/modeling_value_head.py:203]: I think something like that might be needed

**Comment by lvwerra** [comment]
Thanks for working on this @edbeeching! Could you provide a small code snippet on how saving and resuming would work for a user?
